dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/cache/AnalyticsCacheSettings.java

```diff
@@ -0,0 +1,170 @@
+    {
+        final AnalyticsCacheTtlMode analyticsCacheMode = (AnalyticsCacheTtlMode) systemSettingManager
+            .getSystemSetting( ANALYTICS_CACHE_TTL_MODE );
+
+        final CacheStrategy cacheStrategy = (CacheStrategy) systemSettingManager.getSystemSetting( CACHE_STRATEGY );
+
+        return FIXED == analyticsCacheMode && cacheStrategy != null && cacheStrategy.hasExpirationTimeSet();
+    }
+
+    /**
+     * Encapsulates the calculation of the progressive expiration time for the
+     * analytics caching at application level, if the PROGRESSIVE mode is set.
+     * 
+     * @param dateBeforeToday the date to be used during the calculation of the
+     *        progressive expiration time.
+     * 
+     * @return the expiration time computed based on the given "dateBeforeToday".
+     */
+    public long progressiveExpirationTimeOrDefault( final Date dateBeforeToday )
+    {
+        return new TimeToLive( dateBeforeToday, getProgressiveTtlFactorOrDefault() ).compute();
+    }
+
+    /**
+     * Retrieves the expiration time in seconds based on the system settings defined
+     * by the {@link org.hisp.dhis.setting.SettingKey#CACHE_STRATEGY}
+     * 
+     * @see CacheStrategy
+     * 
+     * @return the predefined expiration time set or 0 (ZERO) if nothing is set.
+     */
+    public long fixedExpirationTimeOrDefault()
+    {
+        final CacheStrategy cacheStrategy = (CacheStrategy) systemSettingManager.getSystemSetting( CACHE_STRATEGY );
+
+        if ( cacheStrategy != null && cacheStrategy.hasExpirationTimeSet() )
+        {
+            return cacheStrategy.toSeconds();
+        }
+        else
+        {
+            // Try to get a default value
+            final CacheStrategy defaultExpirationTime = (CacheStrategy) CACHE_STRATEGY.getDefaultValue();
+
+            if ( defaultExpirationTime.hasExpirationTimeSet() )
+            {
+                return defaultExpirationTime.toSeconds();
+            }
+            else
+            {
+                // Return ZERO (always expire)
+                return NO_CACHE.toSeconds();
+            }
+        }
+    }
+
+    /**
+     * Checks if any the caching feature (PROGRESSIVE or FIXED) is enabled.
+     *
+     * @return true if there is any expiration time set, false otherwise.
+     */
+    public boolean isCachingEnabled()
+    {
+        return isFixedCachingEnabled() || isProgressiveCachingEnabled();
+    }
+
+    /**
+     * Returns the TTL factor set in system settings or the default value if nothing
+     * is set.
+     *
+     * If a negative TTL factor was set, the default value will be returned.
+     *
+     * @return the ttl factor
+     */
+    private int getProgressiveTtlFactorOrDefault()
+    {
+        final Integer ttlFactor = (Integer) systemSettingManager
+            .getSystemSetting( ANALYTICS_CACHE_PROGRESSIVE_TTL_FACTOR );
+        final boolean ttlNotNullAndPositive = ttlFactor != null && ttlFactor > 0;
```

Comment:
Here `ttlFactor` cannot be null, as we have a default value for ANALYTICS_CACHE_PROGRESSIVE_TTL_FACTOR. So the null check can be removed.

Then we can do:

```
return Math.max( ttlFactor, 1 )
```

Reply:
Done